File: autotune.py
"""Online PID auto-tuner for the tracking loops.

Tunes the six pid_*_gain values IN PLACE on the live ConfigState while a
target is being tracked. Both control paths (the Python loop and the Rust
core loop) re-read the gains from config every cycle, so writing a candidate
gain takes effect on the next control cycle with no loop-specific wiring --
and the PID pane's sliders visibly follow along.

Method: per-axis coordinate descent ("twiddle") in log-gain space, driven by
the natural tracking excitation (trajectory motion + measurement noise)
rather than injected test signals, so it is safe to run on a live pass:

  * each sweep starts with a measurement window at the current best gains to
    re-baseline the cost (conditions drift over a pass; a stale best-cost
    would block all later acceptance),
  * each parameter (P, then D, then I) is probed up then down by a step in
    decades; a candidate that improves the cost by more than the acceptance
    margin is kept and expands the step, otherwise it is reverted and the
    step shrinks,
  * both axes share the measurement schedule -- their errors are independent
    signals sampled from the same window -- so a full sweep costs 7 windows
    (~40 s at the defaults) instead of 13.

Cost is the RMS mount-axis position error over the window. Guards: gains are
clamped to the UI slider range, a diverging candidate is reverted mid-window,
and the tuner pauses (holding the best-known gains) whenever tracking stops
being live -- STOP, target below the mask, etc. -- resuming the interrupted
probe when tracking returns. A mode change that switches the PID gain
PROFILE (PROGRAM <-> HOTSPOT; see JoystickModeState.service_gain_profiles)
instead STOPS the tune, keeping the best gains: the two modes are different
plants with separately stored tunings, so a tune never outlives its plant.
"""

import math
import time

# Gain clamp: the PID pane's log sliders span exactly this range.
GAIN_MIN = 2.0e-5
GAIN_MAX = 2.0

# Measurement schedule. Settle discards the transient from the gain change;
# eval is the scoring window. At the ~8-15 Hz control cadence a 4 s window
# yields 30-60 error samples.
SETTLE_SEC = 1.5
EVAL_SEC = 4.0
MIN_SAMPLES = 8

# A candidate must beat the best cost by >3% -- windowed RMS is noisy, and a
# plain "<" acceptance random-walks the gains instead of descending.
IMPROVE_FRACTION = 0.97

# Twiddle steps, in decades of gain (multiplicative: 0.3 decades ~= x2).
# Expand on success so a badly-seeded gain crosses its decades quickly;
# shrink on a failed up+down pair. A parameter is converged when its step
# falls under STEP_DONE_DECADES (~ x1.1).
STEP_INIT_DECADES = 0.30
STEP_EXPAND = 1.4
STEP_SHRINK = 0.5
STEP_DONE_DECADES = 0.04
MAX_SWEEPS = 10

# Divergence guard: revert a candidate the moment an axis error exceeds
# max(absolute floor, factor x baseline RMS) -- don't wait out the window.
DIVERGE_ABS_DEG = 5.0
DIVERGE_FACTOR = 4.0

# Probe order: P sets the response, D damps what P excites, I trims bias.
PARAM_ORDER = ('p', 'd', 'i')

_INF = float('inf')


def _clamp(gain):
    return min(GAIN_MAX, max(GAIN_MIN, gain))


class _AxisTuner:
    """Twiddle state for one mount axis (fields pid_<axis>_<p|i|d>_gain)."""

    def __init__(self, config_state, axis):
        self.cfg = config_state
        self.axis = axis
        self.steps = {k: STEP_INIT_DECADES for k in PARAM_ORDER}
        self.initial = self.read_gains()
        self.best = dict(self.initial)
        self.best_cost = None      # windowed RMS (deg) at self.best
        self.baseline_rms = None   # sweep-start RMS, sizes the divergence gate
        self.aborted = False       # current window tripped the divergence gate

    def _field(self, param):
        return f"pid_{self.axis}_{param}_gain"

    def read_gains(self):
        return {k: float(getattr(self.cfg, self._field(k), 0.0) or 0.0)
                for k in PARAM_ORDER}

    def apply(self, gains):
        for k, v in gains.items():
            setattr(self.cfg, self._field(k), round(_clamp(v), 6))

    def candidate(self, param, direction):
        """Best gains with one parameter stepped up (+1) or down (-1)."""
        gains = dict(self.best)
        factor = 10.0 ** self.steps[param]
        base = _clamp(gains[param])
        gains[param] = _clamp(base * factor if direction > 0 else base / factor)
        return gains

    @property
    def done(self):
        return all(s < STEP_DONE_DECADES for s in self.steps.values())

    def diverge_limit(self):
        if self.baseline_rms is None:
            return DIVERGE_ABS_DEG
        return max(DIVERGE_ABS_DEG, DIVERGE_FACTOR * self.baseline_rms)


class PIDAutoTuner:
    """Shared-schedule twiddle over both axes. Drive it by calling update()
    once per control cycle (either loop); read status_text() for the UI."""

    def __init__(self, config_state, mode=None):
        self.cfg = config_state
        # Tracking mode this tuner was armed in. The caller gates samples on
        # it: PROGRAM and HOTSPOT are different plants (encoder loop vs.
        # optical loop), so a tune only ever measures the mode it started in.
        self.armed_mode = mode
        # Set by the caller at arm time: what we're tuning against (target
        # name). Stamped into the mode's gain profile when the tune finishes.
        self.target_label = None
        self._label_stamped = False
        self.axes = {'azm': _AxisTuner(config_state, 'azm'),
                     'alt': _AxisTuner(config_state, 'alt')}
        self.active = False
        self.phase = 'idle'        # idle | settle | measure | paused | done | stopped
        self.stage = None          # ('baseline', 0) or (param, +1/-1)
        self.sweep = 0
        self._param_idx = 0
        self._pending = {}         # axis -> candidate gains under test (None = holding best)
        self._up_accepted = {}     # axis -> up-probe accepted (skip the down-probe)
        self._phase_t0 = 0.0
        self._acc = {}             # axis -> [sum_sq, n]
        self._messages = []

    # ---------------------------------------------------------------- control
    def start(self, now=None):
        """Arm the tuner. `now` is injectable so tests can run a simulated
        clock; the live callers omit it."""
        self.sweep = 0
        self._param_idx = 0
        self.active = True
        self._begin_stage(('baseline', 0), time.time() if now is None else now)
        self._say("Auto-tune started (P/D/I twiddle, both axes)")

    def stop(self, revert=False):
        """Stop tuning. Keeps the best-known gains unless revert=True, which
        restores the gains from before start()."""
        for ax in self.axes.values():
            ax.apply(ax.initial if revert else ax.best)
        self.active = False
        self.phase = 'stopped'
        self._say("Auto-tune stopped - gains "
                  + ("reverted" if revert else f"kept ({self.summary()})"))

    def take_messages(self):
        msgs, self._messages = self._messages, []
        return msgs

    def _say(self, msg):
        print(f"AUTOTUNE: {msg}")
        self._messages.append(msg)

    # ---------------------------------------------------------------- sampling
    def update(self, now, tracking_active, azm_error_deg, alt_error_deg):
        """Feed one control-cycle sample. tracking_active must be True only
        while the armed mode is genuinely closed-loop on the target."""
        if not self.active:
            return

        if not tracking_active:
            if self.phase != 'paused':
                # Hold the best-known gains while paused; the interrupted
                # probe restarts from its settle phase on resume.
                for ax in self.axes.values():
                    ax.apply(ax.best)
                self.phase = 'paused'
                self._say("Auto-tune paused (tracking not live)")
            return

        if self.phase == 'paused':
            self._begin_stage(self.stage, now)
            self._say("Auto-tune resumed")
            return

        errors = {'azm': azm_error_deg, 'alt': alt_error_deg}

        # Divergence gate, live in both settle and measure: a runaway
        # candidate is reverted immediately, not at the end of the window.
        for name, ax in self.axes.items():
            if not ax.aborted and abs(errors[name]) > ax.diverge_limit():
                ax.apply(ax.best)
                ax.aborted = True

        if self.phase == 'settle':
            if now - self._phase_t0 >= SETTLE_SEC:
                self.phase = 'measure'
                self._phase_t0 = now
                self._acc = {name: [0.0, 0] for name in self.axes}
        elif self.phase == 'measure':
            for name, ax in self.axes.items():
                if not ax.aborted:
                    acc = self._acc[name]
                    acc[0] += errors[name] * errors[name]
                    acc[1] += 1
            if now - self._phase_t0 >= EVAL_SEC:
                self._finish_window(now)

    # ---------------------------------------------------------------- stages
    def _begin_stage(self, stage, now):
        self.stage = stage
        self.phase = 'settle'
        self._phase_t0 = now
        self._pending = {}
        kind, direction = stage
        for name, ax in self.axes.items():
            ax.aborted = False
            probe = None
            if kind != 'baseline' and not ax.done:
                if direction > 0:
                    probe = ax.candidate(kind, +1)
                elif not self._up_accepted.get(name, False):
                    probe = ax.candidate(kind, -1)
            ax.apply(probe if probe is not None else ax.best)
            self._pending[name] = probe

    def _finish_window(self, now):
        kind, _direction = self.stage

        # Too few samples with nothing aborted means the loop cadence hiccuped
        # (not the candidate's fault): re-run the same stage.
        if (not any(ax.aborted for ax in self.axes.values())
                and all(acc[1] < MIN_SAMPLES for acc in self._acc.values())):
            self._begin_stage(self.stage, now)
            return

        costs = {}
        for name, ax in self.axes.items():
            sum_sq, n = self._acc[name]
            if ax.aborted or n < MIN_SAMPLES:
                costs[name] = _INF
            else:
                costs[name] = math.sqrt(sum_sq / n)

        if kind == 'baseline':
            for name, ax in self.axes.items():
                if math.isfinite(costs[name]):
                    ax.best_cost = costs[name]
                    ax.baseline_rms = costs[name]
            self._up_accepted = {}
            self._begin_stage((PARAM_ORDER[self._param_idx], +1), now)
            return

        param, direction = kind, _direction
        for name, ax in self.axes.items():
            probe = self._pending.get(name)
            if probe is None:
                continue  # held best gains this window (done, or up accepted)
            accepted = (ax.best_cost is not None
                        and costs[name] < ax.best_cost * IMPROVE_FRACTION)
            if accepted:
                ax.best = probe
                ax.best_cost = costs[name]
                ax.steps[param] = min(2.0, ax.steps[param] * STEP_EXPAND)
                if direction > 0:
                    self._up_accepted[name] = True
            else:
                ax.apply(ax.best)
                if direction < 0:
                    # Both directions failed: tighten this parameter's step.
                    ax.steps[param] *= STEP_SHRINK

        if direction > 0:
            self._begin_stage((param, -1), now)
            return

        # Down-probe finished: advance to the next parameter / sweep.
        self._up_accepted = {}
        self._param_idx += 1
        if self._param_idx < len(PARAM_ORDER):
            self._begin_stage((PARAM_ORDER[self._param_idx], +1), now)
            return
        self._param_idx = 0
        self.sweep += 1
        if all(ax.done for ax in self.axes.values()) or self.sweep >= MAX_SWEEPS:
            for ax in self.axes.values():
                ax.apply(ax.best)
            self.active = False
            self.phase = 'done'
            self._say(f"Auto-tune converged after {self.sweep} sweep(s): {self.summary()}")
            return
        self._begin_stage(('baseline', 0), now)

    # ---------------------------------------------------------------- display
    def summary(self):
        parts = []
        for name, ax in self.axes.items():
            if ax.best_cost is not None:
                parts.append(f'{name} rms {ax.best_cost * 3600.0:.0f}"')
        return ", ".join(parts) if parts else "no measurement yet"

    def status_text(self):
        """One short line for the PID pane."""
        if self.phase == 'paused':
            return "autotune: paused (not tracking)"
        if self.phase == 'done':
            return f"autotune: converged - {self.summary()}"
        if self.phase == 'stopped':
            return f"autotune: stopped - {self.summary()}"
        if self.stage is None:
            return "autotune: idle"
        kind, direction = self.stage
        if kind == 'baseline':
            what = "baseline"
        else:
            what = f"{kind.upper()}{'+' if direction > 0 else '-'}"
        left = ""
        if self.phase == 'measure':
            left = f" {max(0.0, EVAL_SEC - (time.time() - self._phase_t0)):.0f}s"
        return f"autotune S{self.sweep + 1} {what}{left} | {self.summary()}"

File: rust/skytracker-core/src/autotune.rs
//! Online PID auto-tuner — faithful port of autotune.py (Phase 6b of the
//! Rust port): per-axis coordinate descent ("twiddle") in log-gain space
//! on a shared measurement schedule, with settle/eval windows, a live
//! divergence gate, pause/resume, and sweep convergence. The tuner owns
//! the gains it applies; the caller mirrors `applied_gains()` into its
//! live config each cycle (the Python wrapper does exactly that).
//! Parameter order is P, D, I.

pub const GAIN_MIN: f64 = 2.0e-5;
pub const GAIN_MAX: f64 = 2.0;
pub const SETTLE_SEC: f64 = 1.5;
pub const EVAL_SEC: f64 = 4.0;
pub const MIN_SAMPLES: u32 = 8;
pub const IMPROVE_FRACTION: f64 = 0.97;
pub const STEP_INIT_DECADES: f64 = 0.30;
pub const STEP_EXPAND: f64 = 1.4;
pub const STEP_SHRINK: f64 = 0.5;
pub const STEP_DONE_DECADES: f64 = 0.04;
pub const MAX_SWEEPS: u32 = 10;
pub const DIVERGE_ABS_DEG: f64 = 5.0;
pub const DIVERGE_FACTOR: f64 = 4.0;
pub const PARAM_NAMES: [&str; 3] = ["p", "d", "i"];

fn clamp(g: f64) -> f64 {
    g.clamp(GAIN_MIN, GAIN_MAX)
}

/// Python's round(x, 6) for the gain magnitudes in play.
fn round6(x: f64) -> f64 {
    (x * 1e6).round() / 1e6
}

#[derive(Clone, Debug)]
pub struct AxisTuner {
    pub steps: [f64; 3],
    pub initial: [f64; 3],
    pub best: [f64; 3],
    pub applied: [f64; 3],
    pub best_cost: Option<f64>,
    pub baseline_rms: Option<f64>,
    pub aborted: bool,
}

impl AxisTuner {
    fn new(initial: [f64; 3]) -> Self {
        AxisTuner {
            steps: [STEP_INIT_DECADES; 3],
            initial,
            best: initial,
            applied: initial,
            best_cost: None,
            baseline_rms: None,
            aborted: false,
        }
    }

    fn apply(&mut self, gains: [f64; 3]) {
        for k in 0..3 {
            self.applied[k] = round6(clamp(gains[k]));
        }
    }

    fn candidate(&self, param: usize, direction: i32) -> [f64; 3] {
        let mut gains = self.best;
        let factor = 10f64.powf(self.steps[param]);
        let base = clamp(gains[param]);
        gains[param] = clamp(if direction > 0 { base * factor } else { base / factor });
        gains
    }

    fn done(&self) -> bool {
        self.steps.iter().all(|&s| s < STEP_DONE_DECADES)
    }

    fn diverge_limit(&self) -> f64 {
        match self.baseline_rms {
            None => DIVERGE_ABS_DEG,
            Some(b) => DIVERGE_ABS_DEG.max(DIVERGE_FACTOR * b),
        }
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Phase {
    Idle,
    Settle,
    Measure,
    Paused,
    Done,
    Stopped,
}

impl Phase {
    pub fn as_str(&self) -> &'static str {
        match self {
            Phase::Idle => "idle",
            Phase::Settle => "settle",
            Phase::Measure => "measure",
            Phase::Paused => "paused",
            Phase::Done => "done",
            Phase::Stopped => "stopped",
        }
    }
}

/// A probe stage: None = baseline window; Some((param_idx, direction)).
pub type Stage = Option<(usize, i32)>;

pub struct PidAutoTuner {
    pub axes: [AxisTuner; 2], // [azm, alt]
    pub active: bool,
    pub phase: Phase,
    /// None until start(); Some(None) = baseline window.
    pub stage: Option<Stage>,
    pub sweep: u32,
    param_idx: usize,
    pending: [Option<[f64; 3]>; 2],
    up_accepted: [bool; 2],
    phase_t0: f64,
    acc: [(f64, u32); 2],
    messages: Vec<String>,
}

impl PidAutoTuner {
    pub fn new(azm_initial: [f64; 3], alt_initial: [f64; 3]) -> Self {
        PidAutoTuner {
            axes: [AxisTuner::new(azm_initial), AxisTuner::new(alt_initial)],
            active: false,
            phase: Phase::Idle,
            stage: None,
            sweep: 0,
            param_idx: 0,
            pending: [None, None],
            up_accepted: [false, false],
            phase_t0: 0.0,
            acc: [(0.0, 0), (0.0, 0)],
            messages: Vec::new(),
        }
    }

    fn say(&mut self, msg: String) {
        self.messages.push(msg);
    }

    pub fn take_messages(&mut self) -> Vec<String> {
        std::mem::take(&mut self.messages)
    }

    pub fn applied_gains(&self) -> [[f64; 3]; 2] {
        [self.axes[0].applied, self.axes[1].applied]
    }

    pub fn start(&mut self, now: f64) {
        self.sweep = 0;
        self.param_idx = 0;
        self.active = true;
        self.begin_stage(None, now);
        self.say("Auto-tune started (P/D/I twiddle, both axes)".into());
    }

    pub fn stop(&mut self, revert: bool) {
        for ax in self.axes.iter_mut() {
            let g = if revert { ax.initial } else { ax.best };
            ax.apply(g);
        }
        self.active = false;
        self.phase = Phase::Stopped;
        let msg = if revert {
            "Auto-tune stopped - gains reverted".to_string()
        } else {
            format!("Auto-tune stopped - gains kept ({})", self.summary())
        };
        self.say(msg);
    }

    pub fn update(&mut self, now: f64, tracking_active: bool, azm_err: f64, alt_err: f64) {
        if !self.active {
            return;
        }
        if !tracking_active {
            if self.phase != Phase::Paused {
                for ax in self.axes.iter_mut() {
                    let b = ax.best;
                    ax.apply(b);
                }
                self.phase = Phase::Paused;
                self.say("Auto-tune paused (tracking not live)".into());
            }
            return;
        }
        if self.phase == Phase::Paused {
            let stage = self.stage.unwrap_or(None);
            self.begin_stage(stage, now);
            self.say("Auto-tune resumed".into());
            return;
        }

        let errors = [azm_err, alt_err];
        for (i, ax) in self.axes.iter_mut().enumerate() {
            if !ax.aborted && errors[i].abs() > ax.diverge_limit() {
                let b = ax.best;
                ax.apply(b);
                ax.aborted = true;
            }
        }

        match self.phase {
            Phase::Settle => {
                if now - self.phase_t0 >= SETTLE_SEC {
                    self.phase = Phase::Measure;
                    self.phase_t0 = now;
                    self.acc = [(0.0, 0), (0.0, 0)];
                }
            }
            Phase::Measure => {
                for i in 0..2 {
                    if !self.axes[i].aborted {
                        self.acc[i].0 += errors[i] * errors[i];
                        self.acc[i].1 += 1;
                    }
                }
                if now - self.phase_t0 >= EVAL_SEC {
                    self.finish_window(now);
                }
            }
            _ => {}
        }
    }

    fn begin_stage(&mut self, stage: Stage, now: f64) {
        self.stage = Some(stage);
        self.phase = Phase::Settle;
        self.phase_t0 = now;
        self.pending = [None, None];
        for i in 0..2 {
            let up_accepted = self.up_accepted[i];
            let ax = &mut self.axes[i];
            ax.aborted = false;
            let mut probe = None;
            if let Some((param, direction)) = stage {
                if !ax.done() {
                    if direction > 0 {
                        probe = Some(ax.candidate(param, 1));
                    } else if !up_accepted {
                        probe = Some(ax.candidate(param, -1));
                    }
                }
            }
            let g = probe.unwrap_or(ax.best);
            ax.apply(g);
            self.pending[i] = probe;
        }
    }

    fn finish_window(&mut self, now: f64) {
        let stage = self.stage.unwrap_or(None);
        let any_aborted = self.axes.iter().any(|a| a.aborted);
        if !any_aborted && self.acc.iter().all(|a| a.1 < MIN_SAMPLES) {
            self.begin_stage(stage, now);
            return;
        }
        let mut costs = [f64::INFINITY; 2];
        for i in 0..2 {
            let (sum_sq, n) = self.acc[i];
            if !self.axes[i].aborted && n >= MIN_SAMPLES {
                costs[i] = (sum_sq / n as f64).sqrt();
            }
        }

        let Some((param, direction)) = stage else {
            for i in 0..2 {
                if costs[i].is_finite() {
                    self.axes[i].best_cost = Some(costs[i]);
                    self.axes[i].baseline_rms = Some(costs[i]);
                }
            }
            self.up_accepted = [false, false];
            let p = self.param_idx;
            self.begin_stage(Some((p, 1)), now);
            return;
        };

        for i in 0..2 {
            let Some(probe) = self.pending[i] else { continue };
            let ax = &mut self.axes[i];
            let accepted = match ax.best_cost {
                Some(bc) => costs[i] < bc * IMPROVE_FRACTION,
                None => false,
            };
            if accepted {
                ax.best = probe;
                ax.best_cost = Some(costs[i]);
                ax.steps[param] = (ax.steps[param] * STEP_EXPAND).min(2.0);
                if direction > 0 {
                    self.up_accepted[i] = true;
                }
            } else {
                let b = ax.best;
                ax.apply(b);
                if direction < 0 {
                    ax.steps[param] *= STEP_SHRINK;
                }
            }
        }

        if direction > 0 {
            self.begin_stage(Some((param, -1)), now);
            return;
        }

        self.up_accepted = [false, false];
        self.param_idx += 1;
        if self.param_idx < 3 {
            let p = self.param_idx;
            self.begin_stage(Some((p, 1)), now);
            return;
        }
        self.param_idx = 0;
        self.sweep += 1;
        if self.axes.iter().all(|a| a.done()) || self.sweep >= MAX_SWEEPS {
            for ax in self.axes.iter_mut() {
                let b = ax.best;
                ax.apply(b);
            }
            self.active = false;
            self.phase = Phase::Done;
            let msg = format!(
                "Auto-tune converged after {} sweep(s): {}",
                self.sweep,
                self.summary()
            );
            self.say(msg);
            return;
        }
        self.begin_stage(None, now);
    }

    pub fn summary(&self) -> String {
        let mut parts = Vec::new();
        for (name, ax) in ["azm", "alt"].iter().zip(self.axes.iter()) {
            if let Some(c) = ax.best_cost {
                parts.push(format!("{name} rms {:.0}\"", c * 3600.0));
            }
        }
        if parts.is_empty() {
            "no measurement yet".into()
        } else {
            parts.join(", ")
        }
    }

    /// One status line (time-remaining uses the caller's `now`).
    pub fn status_text(&self, now: f64) -> String {
        match self.phase {
            Phase::Paused => return "autotune: paused (not tracking)".into(),
            Phase::Done => return format!("autotune: converged - {}", self.summary()),
            Phase::Stopped => return format!("autotune: stopped - {}", self.summary()),
            _ => {}
        }
        let Some(stage) = self.stage else {
            return "autotune: idle".into();
        };
        let what = match stage {
            None => "baseline".to_string(),
            Some((p, d)) => format!(
                "{}{}",
                PARAM_NAMES[p].to_uppercase(),
                if d > 0 { "+" } else { "-" }
            ),
        };
        let left = if self.phase == Phase::Measure {
            format!(" {:.0}s", (EVAL_SEC - (now - self.phase_t0)).max(0.0))
        } else {
            String::new()
        };
        format!("autotune S{} {}{} | {}", self.sweep + 1, what, left, self.summary())
    }

    pub fn stage_label(&self) -> String {
        match self.stage {
            None => "none".into(),
            Some(None) => "baseline".into(),
            Some(Some((p, d))) => format!("{}{}", PARAM_NAMES[p], if d > 0 { "+" } else { "-" }),
        }
    }
}
